### PR TITLE
Display delegation on members page for Compound

### PR DESCRIPTION
**Pull request by maxwellfoley3**
Implements PR1 and PR2 on this document https://www.notion.so/hicommonwealth/COMP-Delegation-Support-30983db683574e2193fe7000a47bfaf3

Can be demonstrated by viewing the members page on the Marlin community

### CODE REVIEW

File: client/scripts/controllers/chain/ethereum/compound/api.ts [hunk 1]
```diff
@@ -33,8 +32,7 @@ export default class CompoundAPI extends ContractApi<GovernorAlpha> {
         const iface = new utils.Interface(JSON.stringify(ABI));
         const data = iface.encodeFunctionData(tokenName);
         const resultData = await this.Contract.provider.call({ to: this.Contract.address, data });
-        const tokenAddress = utils.getAddress(Buffer.from(utils.stripZeros(resultData)).toString('hex'));
-        this._Token = MPond__factory.connect(tokenAddress, this.Contract.signer || this.Contract.provider);
+        const tokenAddress = utils.getAddress(Buffer.from(utils.stripZeros(resultData)).toString('hex'));        this._Token = MPond__factory.connect(tokenAddress, this.Contract.signer || this.Contract.provider);
```
Comment (zakhap) [inline client/scripts/controllers/chain/ethereum/compound/api.ts:35]: fix this line!

File: client/scripts/views/components/new_proposal_button.ts [hunk 2]
```diff
@@ -67,6 +67,13 @@ export const getNewProposalMenu = (candidates?: Array<[SubstrateAccount, number]
        label: 'New On-Chain Proposal',
        iconLeft: mobile ? Icons.PLUS : undefined,
      }),
+    app.chain?.network === ChainNetwork.Compound && m(MenuItem, {
+      onclick: (e) => navigateToSubpage('/new/proposal/:type', {
+        type: ProposalType.AaveProposal
```
Comment (zakhap) [inline client/scripts/views/components/new_proposal_button.ts:72]: Is Aave the right proposal type?

File: client/scripts/views/pages/members.ts [hunk 3]
```diff
@@ -11,51 +12,119 @@ import PageLoading from 'views/pages/loading';
...
+            if (!Number.isNaN(parseInt(num, 10))) {
+              vnode.state.delegateAmount = num;
+            }
+          },
+        }),
+        m(Button, {
+          label: 'Delegate',
+          intent: 'primary',
+          onclick: async (e) => {
+            (app.chain as Compound).chain.setDelegate(vnode.attrs.address, vnode.state.delegateAmount);
```
Comment (zakhap) [inline client/scripts/views/pages/members.ts:49]: do we need some sort of check here?

File: client/scripts/views/pages/members.ts [hunk 4]
```diff
@@ -11,51 +12,119 @@ import PageLoading from 'views/pages/loading';
...
+      // m(Login),
+      ]),
+    ]);
+  }
+};
+
+const MembersPage : m.Component<{}, { membersRequested: boolean, membersLoaded: MemberInfo[],
+pageToLoad: number, totalMembers: number, isCompound: boolean, voteEvents }> = {
+  oninit: (vnode) => {
+    $.get('/api/getChainEvents?type=marlin-vote-cast').then(({ result }) => {
```
Comment (zakhap) [inline client/scripts/views/pages/members.ts:61]: are all compound "vote-cast" emitted events typed as "marlin-vote-cast"? we should probably change this (and change our defaulting from "MPond" to "Comp"). thoughts @jnaviask?

File: client/scripts/views/pages/members.ts [hunk 5]
```diff
@@ -11,51 +12,119 @@ import PageLoading from 'views/pages/loading';
...
-          const { chain, address } = info;
-          return (!activeMembersHash[`${chain}##${address}`]);
-        });
-        vnode.state.membersLoaded = membersActive.concat(membersInactive).sort((a, b) => b.count - a.count);
-        m.redraw();
+    if (vnode.state.membersRequested) {
+      vnode.state.membersRequested = false;
+      activeInfo.getMembersByPage(activeInfo.id, vnode.state.pageToLoad, MEMBERS_PER_PAGE)
+        .then(({ result, total }) => {
+          const newMembers = result.map(((o, i) => {
```
Comment (zakhap) [inline client/scripts/views/pages/members.ts:110]: remove `i`?

File: client/scripts/views/pages/members.ts [hunk 6]
```diff
@@ -11,51 +12,119 @@ import PageLoading from 'views/pages/loading';
...
+          vnode.state.totalMembers = total;
+
+          const offset = vnode.state.membersLoaded.length - newMembers.length;
+          if (vnode.state.isCompound) {
+            return app.chain.initApi().then(()=>
+              Promise.all(newMembers.map((o, i) => {
+                return (app.chain as Compound).chain.getVotingPower(o.address).then((votes) => {
+                  vnode.state.membersLoaded[offset + i].votes = votes;
+                })
+              })).then(()=>m.redraw())
```
Comment (zakhap) [inline client/scripts/views/pages/members.ts:126]: lint the spacing here and above on 122.

**Review by zakhap — CHANGES_REQUESTED** [review]
Have a few questions: 
1. Why do we need `shared/eth/cache/solidity-files-cache.json`? 
2. I don't see any delegate vote numbers on the Members page, is that because the none of the members we have on commonwealth are delegates or is that a bug? 
3. Do we need the yalc'd chain-events in the repo?

File: client/scripts/views/pages/members.ts (hunk 4 above)
Comment (maxwellfoley3) [reply client/scripts/views/pages/members.ts:68]: Hmm yeah im not sure

**Comment by maxwellfoley3** [comment]
fixed these problems

File: client/scripts/views/pages/members.ts (hunk 4 above)
Comment (jnaviask) [reply client/scripts/views/pages/members.ts:68]: Vote cast events will be emitted as `${chainId}-vote-cast` as with all other chain events.

**Comment by dillchen** [comment]
- [ ] styling changes
- [ ] continued transaction after pushing to staging

File: client/scripts/controllers/chain/ethereum/compound/chain.ts [hunk 7]
```diff
@@ -47,9 +56,9 @@ export default class CompoundChain extends EthereumChain {
         this.compoundApi.Token,
       );
       const gasLimit = await contract.estimateGas.delegate(address, amount);
+      console.log(this.balanceOf(address));
```
Comment (dillchen) [inline client/scripts/controllers/chain/ethereum/compound/chain.ts:59]: Remove, but also adjust:     

const votesByDecimals = votes.div(BigNumber.from(10).pow(BigNumber.from(this.compoundApi.decimals)));

**Comment by dillchen** [comment]
@zakhap one thing left on this is that sometimes the members page delegate and voting power columns disappear after a while `vnode.state.isCompound` not a good way to set this

**Comment by zakhap** [comment]
@jnaviask @gdjohnson @dillchen I think this is finally ready to be merged, but because there were so many QA bugs I fixed, I'd love another QA pass from one of you.

Things to check:
1. Compound-based Members Page (Marlin/members)
2. Normal Members Page (Edgeware/members)
3. No more flashing on page (both types above), esp. on scroll+pagination
4. Pagination works?

**Comment by gdjohnson** [comment]
Would be good to get sorting (by post count, funds) added. Kinda weird having a table that can't be sorted. Since we already have all the data, shouldn't be so hard—pretty much involves adding a simple boolean sortBy switch, a .sort method—maybe an hour tops? 30 min? If this is of interest @zakhap @dillchen  I can implement now or make a ticket for later. 

But more importantly, no flashing on my end, Edgeware + Marlins page look good, infinite scroll seems to work.

However, members pages on offchain communities are totally broken. Community pages are getting passed as if they were chain ids... Writing fix now.

**Comment by gdjohnson** [comment]
Things working nicely for me—tried some offchain communities, some compound-based chain comms, and some regular chain comms. Made some improvements to spinner handling and swapped out some variable names for legibility/clarity.

There are some issues with mobile/tablet UI—see attached images. Any ideas what we wanna do about this?
image
image

**Comment by jnaviask** [comment]
Reopening this because I need to revert.

File: client/scripts/controllers/chain/ethereum/aave/api.ts [hunk 8]
```diff
@@ -41,6 +44,8 @@ export default class AaveApi extends ContractApi<IAaveGovernanceV2> {
     await super.init();
     this._Governance = this.Contract;
 
+    await this._Token.decimals().then((d) => this._decimals = d);
```
Comment (jnaviask) [inline client/scripts/controllers/chain/ethereum/aave/api.ts:47]: We have no guarantee that `_Token` is initialized at this point in time.

File: client/scripts/helpers/index.ts [hunk 9]
```diff
@@ -227,6 +229,20 @@ export function formatNumberLong(num: number) {
   return nf.format(num);
 }
 
+// export function formatBNLong(num: BigNumber) {
```
Comment (jnaviask) [inline client/scripts/helpers/index.ts:232]: Dead code?

File: server/routes/bulkMembers.ts [hunk 10]
```diff
@@ -4,17 +4,49 @@ import { factory, formatFilename } from '../../shared/logging';
...
+        { model: models.Address, as: 'Address' },
+        'last_active',
+        'DESC NULLS LAST',
+      ],
+    ],
+    offset:
+      req.query.itemsPerPage && req.query.pageNum
+        ? req.query.itemsPerPage * req.query.pageNum
+        : undefined,
+    limit: req.query.itemsPerPage,
```
Comment (jnaviask) [inline server/routes/bulkMembers.ts:42]: What do these query changes do?

File: server/routes/getChainEvents.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import { Request, Response, NextFunction } from 'express';
+import { sequelize, DB } from '../database';
+
+const getChainEvents = async (
```
Comment (jnaviask) [inline server/routes/getChainEvents.ts:4]: Why do we need this route?

File: client/scripts/controllers/chain/ethereum/aave/api.ts (hunk 8 above)
Comment (zakhap) [reply client/scripts/controllers/chain/ethereum/aave/api.ts:47]: Removed for the time-being, this is used for member's delegation page which we don't have enabled yet for Aave.

File: client/scripts/helpers/index.ts (hunk 9 above)
Comment (zakhap) [reply client/scripts/helpers/index.ts:232]: Removed

File: server/routes/bulkMembers.ts (hunk 10 above)
Comment (zakhap) [reply server/routes/bulkMembers.ts:42]: These are from @gdjohnson's PR that fixed the pagination for offchain-community fetching.

File: server/routes/getChainEvents.ts (hunk 11 above)
Comment (zakhap) [reply server/routes/getChainEvents.ts:4]: This is how we get the vote-cast events to calculate pagination for which members to load on delegation pages first.

**Pull request merged**